File: ios/Classes/Permission.swift
//
//  Permission.swift
//  get_permission
//
//  Created by Yuri on 18.12.2021.
//

import Foundation

enum PermissionType: Int {
    case camera = 0
    case microphone
    case contacts
    case notification
    case notificationOptions
    case notificationOptionAlert
    case notificationOptionBadge
    case notificationOptionSound
    case notificationOptionCarPlay
    case notificationOptionCriticalAlert
    case notificationOptionProvisional
    case notificationOptionAnnouncement
    case notificationOptionTimeSensitive
    case appTrackingTransparency
    case calendar
    case reminder
    case speech
    case locationAlways
    case locationWhenInUse
    case photos
    case photosWriteOnlyIOS
    case mediaLibrary
    case phone
}

enum NotificationOption: Int {
    case alert
    case badge
    case sound
    case carPlay
    case criticalAlert
    case provisional
    case announcement
    case timeSensitive
}

enum PermissionStatus: Int {
    case denied = 0
    case authorized
    case restricted
    case limited
    case permanentlyDenied // Not used on iOS
    case notSupported // In case of OS version does not supports permission
}

enum PermissionAvailability: Int {
    case enabled = 0
    case disabled
    case nonApplicable
}

File: lib/permission.dart
import 'dart:async';

abstract class PermissionHandler {
  static Future<Status> checkStatus(Permissions permission) async {
    throw UnimplementedError('Not implemented');
  }

  static Future<Availability> checkAvailability(Permissions permission) async {
    throw UnimplementedError('Not implemented');
  }

  static Future<Status> request(Permissions permission) async {
    throw UnimplementedError('Not implemented');
  }
}

enum Permissions {
  camera,
  microphone,
  contacts,

  /// The notification permission with options: alert, badge, sound
  notification,

  /// The notification permission with any options
  notificationOptions,
  notificationOptionAlert,
  notificationOptionBadge,
  notificationOptionSound,
  notificationOptionCarPlay,

  /// To use Critical Alert permission the app should have
  /// a granted entitlement by Apple: com.apple.developer.usernotifications.critical-alerts
  notificationOptionCriticalAlert,
  notificationOptionProvisional,
  notificationOptionAnnouncement,
  notificationOptionTimeSensitive,
  appTrackingTransparencyIOS,

  calendar,
  reminderIOS,

  speech,

  locationAlways,
  locationWhenInUse,

  photos,
  photosWriteOnlyIOS,

  mediaLibrary,

  phoneAndroid,

  sensors,
  smsAndroid,
  storage,
  ignoreBatteryOptimizationsAndroid,
  mediaLocation,
  activityRecognitionAndroid,
  bluetooth,
  manageExternalStorageAndroid,
  systemAlertWindowAndroid,
  requestInstallPackageAndroid,
  accessNotificationPolicyAndroid,
  bluetoothScanAndroid,
  bluetoothAdvertiseAndroid,
  bluetoothConnectAndroid,
  unknown,
}

class PermissionOption {
  final int value;
  PermissionOption(this.value);

  static final notificationOptionAlert =
      PermissionOption(NotificationOption.alert.index);
  static final notificationOptionBadge =
      PermissionOption(NotificationOption.badge.index);
  static final notificationOptionSound =
      PermissionOption(NotificationOption.sound.index);
  static final notificationOptionCarPlay =
      PermissionOption(NotificationOption.carPlay.index);
  static final notificationOptionCriticalAlert =
      PermissionOption(NotificationOption.criticalAlert.index);
  static final notificationOptionAnnouncement =
      PermissionOption(NotificationOption.announcement.index);
  static final notificationOptionTimeSensitive =
      PermissionOption(NotificationOption.timeSensitive.index);
}

enum NotificationOption {
  alert,
  badge,
  sound,
  carPlay,
  criticalAlert,
  provisional,
  announcement,
  timeSensitive,
}

enum Status {
  denied,
  authorized,
  restrictedIOS,
  limitedIOS,
  permanentlyDeniedAndroid,
  notSupported, // In case of OS version does not supports permission
}

enum Availability {
  enabled,
  disabled,
  nonApplicable,
}

extension AvailabilityParser on Availability {
  static Availability statusFrom(int value) {
    return [
      Availability.enabled,
      Availability.disabled,
      Availability.nonApplicable,
    ][value];
  }
}
